LPS-130112 Pass contents as a prop and avoid unwanted onChange

julien opened the pull request:
Currently, when using `ClassicEditor`, the `onChange` callback is called
when the editor is "ready", because we're calling `setData` in
the `onInstanceReady` callback.  Ideally, the `onChange` callback
should be called when the editor's contents have changed,
so to fix that, we're passing the content as the `data` prop.

**Before**

![](https://user-images.githubusercontent.com/5572/113835243-134d9800-978c-11eb-949c-8cc3d7a88f5c.gif)

**After**

![](https://user-images.githubusercontent.com/5572/113835261-1a74a600-978c-11eb-8e5d-d08d9ae217d0.gif)

**Test plan**:
- Navigate to **Content & Data** > **Web Content**
- Create a new **Basic Web Content** and publish it
- Put a breakpoint (or a logpoint) [here](https://git.io/JYbSQ)
- Edit the web content you created previously
- Assert that the breakpoint is not "hit" when the Web Content's editor
  is initalized and ready

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

julien commented:
cc @boton

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-16.liferay.com/job/test-portal-source-format/5203/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-16.liferay.com//userContent/jobs/test-portal-source-format/builds/5203/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 3 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7">9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/julien/liferay-portal/tree/LPS-130112">LPS-130112</a><br/>Branch GIT ID: <a href="https://github.com/julien/liferay-portal/commit/316c3ba0dcb242db03ff7506754e39115eeda4a7">316c3ba0dcb242db03ff7506754e39115eeda4a7</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-16.liferay.com/job/test-portal-source-format/5203/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-16.liferay.com/userContent/jobs/test-portal-source-format/builds/5203/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-16.liferay.com/job/test-portal-source-format/5203/">test-portal-source-format#5203</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-16.liferay.com/userContent/jobs/test-portal-source-format/builds/5203/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/954">liferay-frontend#954</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1523944087">[master] ci:test:sf - julien &gt; liferay-frontend - PR#954 - 2021-04-07[01:34:42]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-16.liferay.com/job/publish-testray-report/137/">publish-testray-report#137</a><br/></div></div>

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>relevant</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-16.liferay.com/job/test-portal-acceptance-pullrequest(master)/8111/">test-portal-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-16.liferay.com//userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8111/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:x: ci:test:relevant - 0 out of 1 jobs passed in 3 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7">9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7</a></p><h4>Upstream Comparison:</h4><p>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7">9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7</a><br/>Jenkins Build URL: <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)//1732">Acceptance Upstream DXP (master) #1732</a></p><details><summary><strong>ci:test:relevant - 0 out of 1 jobs PASSED</strong></summary><div><h4>1 Failed Jobs:</h4><ul><li><a href="https://test-1-16.liferay.com/job/test-portal-acceptance-pullrequest(master)/8111/"><strike><strong>test-portal-acceptance-pullrequest(master)</strong></strike></a></li></ul></div></details><h5>For more details click <a href="https://test-1-16.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8111/jenkins-report.html">here</a>.</h5><hr/><h4>Failures unique to this pull:</h4><ol><li><div><h5><a href="https://test-1-16.liferay.com/job/test-portal-acceptance-pullrequest(master)/8111/">test-portal-acceptance-pullrequest(master)</a></h5><div><h6>Job Results:</h6><p>0 Jobs Passed.<br/>1 Job Failed.</p></div><pre><code>	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
	at org.apache.tools.ant.Task.perform(Task.java:350)
	at java.util.Vector.forEach(Vector.java:1249)
	at org.apache.tools.ant.taskdefs.Sequential.execute(Sequential.java:67)
	at net.sf.antcontrib.logic.IfTask.execute(IfTask.java:217)
	at sun.reflect.GeneratedMethodAccessor29.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
	at org.apache.tools.ant.TaskAdapter.execute(TaskAdapter.java:155)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
	at org.apache.tools.ant.Task.perform(Task.java:350)
	at java.util.Vector.forEach(Vector.java:1249)
	at org.apache.tools.ant.taskdefs.Sequential.execute(Sequential.java:67)
	at net.sf.antcontrib.logic.TryCatchTask\$CatchBlock.execute(TryCatchTask.java:137)
	at net.sf.antcontrib.logic.TryCatchTask.execute(TryCatchTask.java:225)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
	at org.apache.tools.ant.Task.perform(Task.java:350)
	at org.apache.tools.ant.Target.execute(Target.java:449)
	at org.apache.tools.ant.Target.performTasks(Target.java:470)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1401)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1374)
	at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
	at org.apache.tools.ant.Project.executeTargets(Project.java:1264)
	at org.apache.tools.ant.Main.runBuild(Main.java:827)
	at org.apache.tools.ant.Main.startAnt(Main.java:223)
	at org.apache.tools.ant.launch.Launcher.run(Launcher.java:280)
	at org.apache.tools.ant.launch.Launcher.main(Launcher.java:109)</code></pre></div></li></ol><h4>For upstream results, click <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)">here</a>.</h4></details></html><h5>Test bundle downloads:</h5><ul><li><a href="https://files.liferay.com/private/bundles/test-1-16/jobs/test-portal-acceptance-pullrequest(master)/builds/8111/liferay-portal-bundle-tomcat.tar.gz">liferay-portal-bundle-tomcat.tar.gz</a> (mirrors: <a href="http://mirrors.dlc.liferay.com/files.liferay.com/private/bundles/test-1-16/jobs/test-portal-acceptance-pullrequest(master)/builds/8111/liferay-portal-bundle-tomcat.tar.gz">dlc</a>, <a href="http://mirrors.lax.liferay.com/files.liferay.com/private/bundles/test-1-16/jobs/test-portal-acceptance-pullrequest(master)/builds/8111/liferay-portal-bundle-tomcat.tar.gz">lax</a>)</li></ul>

jbalsas reviewed (commented):
The problem with this approach is that content is first placed in the page and then into the editor, so you can see a very weird-looking flash of content.

I think this is a very visible issue that we should try to avoid and probably the reason why we set the data `onInstanceReady`.

This should also be even more visible if we have processors like bbcode or creole that completely change the html markup, so it would be good to test there as well.

jbalsas commented:
On top of that, it's a well documented [behaviour of the `change` event](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-change) in CKEditor that:

> Due to performance reasons, it is not verified if the content really changed. The editor instead watches several editing actions that usually result in changes. This event may thus in some cases be fired when no changes happen or may even get fired twice.
>
> If it is important not to get the change event fired too often, you should compare the previous and the current editor content inside the event listener. It is not recommended to do that on every change event.

boton commented:
Maybe if we need to avoid the flash of content, we try to use the other workaround that @julien mentioned yesterday: use the `internal`  option.

https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#method-setData

Something like this:

```js
onInstanceReady={({editor}) => {
    editor.setData(contents, {internal: true});
}}
```

But I´m not sure is if enough with this in mind 

> On top of that, it's a well documented behaviour of the change event in CKEditor ...

julien commented:
@jbalsas @boton let me have a look into that and see how I can improve it. We definitely don't want a flash of content and it would be nice to avoid those unwanted ``onChange` callbacks

julien commented:
Since it was quick to test, I just wanted to share that passing `{internal: true}` as the second argument to `editor.setData` avoids that `onChange` callback on "initialization", and this is how it looks like:

![](https://user-images.githubusercontent.com/5572/113874687-57ee2900-97b6-11eb-8a8c-d69c231fa202.gif)

You'll see that when the editor is "focussed" there does seem to have a visual "glitch" because the height of the editor **does** change. 

On the other hand, I have a feeling that this is already happening now, or at least this is what I see when I edit an existing Web Content (have a look at the editor, and another look just at the console: I see that the editor's height changes and two "onChange" messages in the console)

![](https://user-images.githubusercontent.com/5572/113875509-1ca02a00-97b7-11eb-92fc-f416d7a31ccd.gif)

I think using `editor.setData(contents, {internal: true})` on the [`onInstanceReady`](https://github.com/liferay/liferay-portal/blob/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L157-L159) callback doesn't look risky after looking at the [`setData`](https://github.com/ckeditor/ckeditor4/blob/17a1555f7fbb0e83f737c9efe27650dca8dff36f/core/editor.js#L1125-L1157) function, but I'll try to find more information just to make sure.

julien commented:
ci:test:sf

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-7.liferay.com/job/test-portal-source-format/4017/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-7.liferay.com//userContent/jobs/test-portal-source-format/builds/4017/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:x: ci:test:sf - 0 out of 1 jobs passed in 1 minute</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7">9c07f7bca2bbde25a8cbcf2c84b6b311fb60e0f7</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/julien/liferay-portal/tree/LPS-130112">LPS-130112</a><br/>Branch GIT ID: <a href="https://github.com/julien/liferay-portal/commit/84432f7ed93f4c474f22b44f48548f1ba6769dd7">84432f7ed93f4c474f22b44f48548f1ba6769dd7</a></p>0 out of 1jobs PASSED<div><h4>1 Failed Jobs:</h4><ul><li><a href="https://test-1-7.liferay.com/job/test-portal-source-format/4017/"><strike><strong>test-portal-source-format</strong></strike></a></li></ul></div><h5>For more details click <a href="https://test-1-7.liferay.com/userContent/jobs/test-portal-source-format/builds/4017/jenkins-report.html">here</a>.</h5><pre><code>	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
	at org.apache.tools.ant.Task.perform(Task.java:350)
	at org.apache.tools.ant.taskdefs.MacroInstance.execute(MacroInstance.java:391)
	... 98 more
Caused by: Sourced file: inline evaluation of: `` 								 									import com.liferay.jenkins.results.parser.GitHubDevSyncUtil;  . . . '' : TargetError : at Line: 91 : in file: inline evaluation of: `` 								 									import com.liferay.jenkins.results.parser.GitHubDevSyncUtil;  . . . '' : throw e ; 

Target exception: java.lang.RuntimeException: java.net.NoRouteToHostException: No route to host (Host unreachable)

	at bsh.BSHThrowStatement.eval(Unknown Source)
	at bsh.BSHBlock.evalBlock(Unknown Source)
	at bsh.BSHBlock.eval(Unknown Source)
	at bsh.BSHBlock.eval(Unknown Source)
	at bsh.BSHTryStatement.eval(Unknown Source)
	at bsh.Interpreter.eval(Unknown Source)
	at bsh.Interpreter.eval(Unknown Source)
	at bsh.Interpreter.eval(Unknown Source)
	at com.liferay.ant.beanshell.BeanShellTask.execute(BeanShellTask.java:65)
	... 141 more
[beanshell] Created cache file in /tmp/jenkins-cached-files/827968559.txt

BUILD FAILED
/opt/dev/projects/github/liferay-jenkins-ee/commands/build-common.xml:12823: The following error occurred while executing this line:
/opt/dev/projects/github/liferay-jenkins-ee/commands/build-test-portal-source-format.xml:18: Sourced file: inline evaluation of: `` 			 				import com.liferay.jenkins.results.parser.AntUtil; 				import com.lifer . . . '' : TargetError : at Line: 345 : in file: inline evaluation of: `` 			 				import com.liferay.jenkins.results.parser.AntUtil; 				import com.lifer . . . '' : throw e ; 

Target exception: The following error occurred while executing this line:
/opt/dev/projects/github/liferay-jenkins-ee/commands/build-common.xml:9685: Sourced file: inline evaluation of: `` 								 									import com.liferay.jenkins.results.parser.GitHubDevSyncUtil;  . . . '' : TargetError : at Line: 91 : in file: inline evaluation of: `` 								 									import com.liferay.jenkins.results.parser.GitHubDevSyncUtil;  . . . '' : throw e ; 

Target exception: java.lang.RuntimeException: java.net.NoRouteToHostException: No route to host (Host unreachable)</code></pre></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-7.liferay.com/job/test-portal-source-format/4017/">test-portal-source-format#4017</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-7.liferay.com/userContent/jobs/test-portal-source-format/builds/4017/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/954">liferay-frontend#954</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1523978876">[master] ci:test:sf - julien &gt; liferay-frontend - PR#954 - 2021-04-07[08:34:10]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-7.liferay.com/job/publish-testray-report/202/">publish-testray-report#202</a><br/></div></div>

boton commented:
ci:test:sf

boton commented:
ci:test:relevant

julien commented:
ci:close

julien commented:
ci:test:sf

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-2.liferay.com/job/test-portal-source-format/309/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-2.liferay.com//userContent/jobs/test-portal-source-format/builds/309/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 3 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/245d7f1752e0f05da0949b93ec60e181e0cde39c">245d7f1752e0f05da0949b93ec60e181e0cde39c</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/julien/liferay-portal/tree/LPS-130112">LPS-130112</a><br/>Branch GIT ID: <a href="https://github.com/julien/liferay-portal/commit/84432f7ed93f4c474f22b44f48548f1ba6769dd7">84432f7ed93f4c474f22b44f48548f1ba6769dd7</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-2.liferay.com/job/test-portal-source-format/309/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-2.liferay.com/userContent/jobs/test-portal-source-format/builds/309/jenkins-report.html">here</a>.</h5></details></html>

julien commented:
> ci:test:sf - 1 out of 1 jobs passed in 3 minutes

abril ... cerral and it magically works :joy:

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-2.liferay.com/job/test-portal-source-format/309/">test-portal-source-format#309</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-2.liferay.com/userContent/jobs/test-portal-source-format/builds/309/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/954">liferay-frontend#954</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1524206779">[master] ci:test:sf - julien &gt; liferay-frontend - PR#954 - 2021-04-08[08:37:10]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-2.liferay.com/job/publish-testray-report/275/">publish-testray-report#275</a><br/></div></div>

julien commented:
ci:test:relevant

jbalsas commented:
> abril ... cerral and it magically works 😂

Hahaha... you'll have to explain that in english and provide video evidence of where that joke comes from!! 😉

julien commented:
@jbalsas I'll try to do my best, promised, this is something I discovered very recently (as you might have guessed), and I thought it was the perfect time to use it.

For the english explanation, it's going to take me some time but I'll work on it :joy:

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>relevant</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest(master)/9419/">test-portal-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-3.liferay.com//userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/9419/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:stable - 9 out of 9 jobs passed</h3><h3>:heavy_check_mark: ci:test:relevant - 23 out of 23 jobs passed in 3 hours 34 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/245d7f1752e0f05da0949b93ec60e181e0cde39c">245d7f1752e0f05da0949b93ec60e181e0cde39c</a></p><h4>Upstream Comparison:</h4><p>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/245d7f1752e0f05da0949b93ec60e181e0cde39c">245d7f1752e0f05da0949b93ec60e181e0cde39c</a><br/>Jenkins Build URL: <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)//1739">Acceptance Upstream DXP (master) #1739</a></p><details><summary><strong>ci:test:stable - 9 out of 9 jobs PASSED</strong></summary><details><summary><strong>9 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6609/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326954/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6615/">test-portal-acceptance-pullrequest-batch(master)/js-unit-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326955/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326952/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326956/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326953/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6610/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/465660/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><details><summary><strong>ci:test:relevant - 23 out of 23 jobs PASSED</strong></summary><details><summary><strong>23 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest(master)/9419/">test-portal-acceptance-pullrequest(master)</a></li><li><a href="https://test-1-21.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/410146/">test-portal-acceptance-pullrequest-batch(master)/central-requirements-jdk8/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6613/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk11_open/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6609/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-13.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430900/">test-portal-acceptance-pullrequest-batch(master)/functional-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/465663/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326954/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6615/">test-portal-acceptance-pullrequest-batch(master)/js-unit-jdk8/0</a></li><li><a href="https://test-1-13.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430901/">test-portal-acceptance-pullrequest-batch(master)/lpkg-base-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/465661/">test-portal-acceptance-pullrequest-batch(master)/lpkg-controller-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/465659/">test-portal-acceptance-pullrequest-batch(master)/modules-compile-jdk8/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6614/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326955/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6612/">test-portal-acceptance-pullrequest-batch(master)/modules-semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/465662/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326952/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6611/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-project-templates-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326956/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-22.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/326953/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-21.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/410147/">test-portal-acceptance-pullrequest-batch(master)/semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6610/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-2.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/6608/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/465660/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><h5>For more details click <a href="https://test-1-3.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/9419/jenkins-report.html">here</a>.</h5></details></html><h5>Test bundle downloads:</h5><ul><li><a href="https://files.liferay.com/private/bundles/test-1-3/jobs/test-portal-acceptance-pullrequest(master)/builds/9419/liferay-portal-bundle-tomcat.tar.gz">liferay-portal-bundle-tomcat.tar.gz</a> (mirrors: <a href="http://mirrors.dlc.liferay.com/files.liferay.com/private/bundles/test-1-3/jobs/test-portal-acceptance-pullrequest(master)/builds/9419/liferay-portal-bundle-tomcat.tar.gz">dlc</a>, <a href="http://mirrors.lax.liferay.com/files.liferay.com/private/bundles/test-1-3/jobs/test-portal-acceptance-pullrequest(master)/builds/9419/liferay-portal-bundle-tomcat.tar.gz">lax</a>)</li></ul>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest(master)/9419/">test-portal-acceptance-pullrequest(master)#9419</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-3.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/9419/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>relevant<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/954">liferay-frontend#954</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1524301780">[master] ci:test:relevant - julien &gt; liferay-frontend - PR#954 - 2021-04-08[10:45:52]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-3.liferay.com/job/publish-testray-report/267/">publish-testray-report#267</a><br/></div></div>

julien commented:
ci:forward

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

The pull request will automatically be forwarded to the user `brianchandotcom` if the following test suites pass:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**stable**

liferay-continuous-integration commented:
Skipping previously passed test suites:
`ci:test:relevant`
`ci:test:sf`

liferay-continuous-integration commented:
All required test suite(s) passed.
Forwarding pull request to `brianchandotcom`.
[Console](https://test-1-6.liferay.com/job/forward-pullrequest/230//consoleFull)

liferay-continuous-integration commented:
Pull request has been successfully forwarded to https://github.com/brianchandotcom/liferay-portal/pull/100662
[Console](https://test-1-6.liferay.com/job/forward-pullrequest/230//consoleFull)

julien commented:
I'm reopening this pull request, because I think I didn't take the right decision here.

In my first change, I suggested using the `data` prop to set the editor's content, because that's what the `CKEditor` React component [exposes](https://github.com/ckeditor/ckeditor4-react/blob/01e0245ef608d5527d2e7510300ed9ce4fc37a14/src/ckeditor.jsx#L143) and I *still* think that's what we should use. I "force pushed" on this branch to avoid multiple commits, but that change is still visible [here](https://github.com/liferay-frontend/liferay-portal/commit/316c3ba0dcb242db03ff7506754e39115eeda4a7).

Additionally, I think there's no need for our [`onInstanceReady`](https://github.com/liferay/liferay-portal/blob/9275a579449189691160bf2dbf5786fd64d1afab/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L157) callback, because:

1. Internally, CKEditor is going to do exactly [the same thing](https://github.com/ckeditor/ckeditor4-react/blob/01e0245ef608d5527d2e7510300ed9ce4fc37a14/src/ckeditor.jsx#L89-L95) (well, to be fair not exactly, because since I have now changed this, [we're passing `{internal: true}` as a second argument to `editor.setData()`](https://github.com/liferay/liferay-portal/blob/9275a579449189691160bf2dbf5786fd64d1afab/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L158) and *they* are passing a `priority` argument with a value of `-1` to [`editor.on`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#method-on)), so this feels like we're adding something that's not needed.

2. Someone using `ClassicEditor` could still pass an `onInstanceReady` prop that [would be passed to `Editor`](https://github.com/liferay/liferay-portal/blob/9275a579449189691160bf2dbf5786fd64d1afab/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L161) and [CKEditor](https://github.com/liferay/liferay-portal/blob/9275a579449189691160bf2dbf5786fd64d1afab/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js#L32), overriding our "default" behavior (the same way some could also pass `data`, thinking that's what supposed to be used, and be surprised of the component's behavior).

3. I can't remember why I did [this](https://github.com/liferay-frontend/liferay-portal/pull/609/files#diff-d275711f5231410d438e4b188e62c43e95d508573bf135dbd9b5ecbe36bbda34R139) at one point (I think it was to use [CKEditor "special hidden" props](https://github.com/ckeditor/ckeditor4-react/blob/01e0245ef608d5527d2e7510300ed9ce4fc37a14/src/ckeditor.jsx#L105-L115) instead of having an event listener in our code, although it's basically "the same thing" - seems like a "nasty" pattern for React components btw): I just know that [initially](https://github.com/liferay/liferay-portal/blob/bfb7c8bd8902f02df8bb8413b8e6a2168b6c2d7f/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L110) we were using the `data` prop and  later we changed it to use the `instanceReady` event handler (but not the `onInstanceReady` prop) [here](https://github.com/liferay/liferay-portal/blob/c6eb4ff077a6436b6f9297566aca66fe7c09c865/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js) for [LPS-113302](https://issues.liferay.com/browse/LPS-113302): the issue doesn't say much but the commit has some information:
 
 > LPS-113302 Set data after instance is ready. That way plugins are applied before setting data, including encoding plugins like bbcode.
 
 That still sounds a bit "vague" for me, maybe it's more a problem with our `bbcode` plugin ... and in anycase since the CKEditor React component is setting the data when the `instanceReady` event is dispatched, I don't understand the reason, so I'll have to investigate. 
 

The only doubt I still have is what @jbalsas commented [here](https://github.com/liferay-frontend/liferay-portal/pull/954#pullrequestreview-629834485)

> The problem with this approach is that content is first placed in the page and then into the editor, so you can see a very weird-looking flash of content.

I've had a look at this several times, and here's a .gif of what I see 
(*Please note that I'm refreshing the page several times to try and see the flash of content*)

![recording](https://user-images.githubusercontent.com/5572/114707651-c25d1700-9d2a-11eb-92fd-fbfd81a9a4f3.gif)

I also used Chrome's devtools to change the "Throttling" to **Fast 3G** but at that point loading the JavaScript was failing so I wasn't really able to see anything or to capture a .gif.

The only "flash of content" I do see is when I focus inside the editor, as can been seen in this other .gif 
(*Pay attention to the editor in which the user types the content, and you'll  see that its height changes*) 

![recording (1)](https://user-images.githubusercontent.com/5572/114708367-8bd3cc00-9d2b-11eb-8d16-26db6c080eb9.gif)

I removed the `onInstanceReady` callback and used the `data` prop for the content, but that doesn't change anything to this behavior - so that still needs more investigation.

Anyway, there's nothing urgent about this but I still would like to leave this discussion open to know what you all think, my main point being that if we can simplify things it's always nicer.

julien commented:
Converting to draft to avoid an accidental forward, but still want this to be discussed

julien commented:
I also tested creating, editing and replying to threads in **Content & Data** > **Message Boards** 
(related to my comment above and [LPS-113302](https://issues.liferay.com/browse/LPS-113302)), and using the `data` prop (+ removing the `onInstanceReady` prop) doesn't cause any problems.

boton commented:
Hello there, after this fix of the editor.setData(contents, {internal: true}) the initial values aren't set on my implementation. Before the fix onChange is called and shows a warning to the author about some unsaved changes.

![image](https://user-images.githubusercontent.com/67901/114733450-a154f000-9d43-11eb-8b1f-983e36caf76e.png)
https://github.com/liferay-lima/liferay-portal/pull/1641#issuecomment-806019526 

I thought I had tested correctly both solutions the `data` and the `editor.setData(contents, {internal: true})`, but it wasn't the case 😢 

> Someone using ClassicEditor could still pass an onInstanceReady prop that would be passed to Editor and CKEditor, overriding our "default" behavior (the same way some could also pass data, thinking that's what supposed to be used, and be surprised of the component's behavior).

Using the `data` prop directly on ClassicEditor fixes my problem.

boton commented:
I´m actually using data instead of content https://github.com/liferay-lima/liferay-portal/pull/1766/commits/3ed3690cc373569598f609f77da84f38fb527000 because the content doesn´t get the initial values 😢

julien commented:
@boton thanks for the additional information.
I think that right now not be the best time to change this, because of the release. But I agree that it's still an important thing we should discuss and solve. Once the release is cut, I can prepare a PR with the change and our motivations to push this forward. Does that seem reasonable to you?

boton commented:
@julien is totally reasonable to me, I'm with you right now it isn't the best time to change this. Ty!

julien commented:
I'll just add yet another comment, to show my interest :hugs: .

Maybe it wasn't very visible in my [previous comment](https://github.com/liferay-frontend/liferay-portal/pull/954#issuecomment-819491906), but the `CKEditor` React component is already doing something [very similar](https://github.com/ckeditor/ckeditor4-react/blob/c0c58bdb26d9b3cf401971e0cf4305a46a2a6fda/src/ckeditor.jsx#L89-L95) than what we're doing [here](https://github.com/liferay/liferay-portal/blob/6bcd6f83bd8bcc185830b6d87d267a217df3c2e1/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L158), so this give me the feeling that we're doing the same thing twice, which feels unnecessary.

boton commented:
In with you @julien as I said in my previous comment https://github.com/liferay-frontend/liferay-portal/pull/954#issuecomment-823966426 I need to use data instead of content to get the initial values https://github.com/liferay/liferay-portal/commit/3ed3690cc373569598f609f77da84f38fb527000.

markocikos commented:
Followup in https://github.com/liferay-frontend/liferay-portal/pull/1133
